### PR TITLE
[maskedtensor] adding unary and binary operations

### PR DESCRIPTION
**Pull request by george-qi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82841
* #82839
* __->__ #82837
* #82836

### CODE REVIEW

File: torch/masked/maskedtensor/binary.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        result_data = torch.sparse_csr_tensor(crow, col, v)
+
+    else:
+        result_data = fn(*data_args)
+
+    if inplace:
+        args[0]._set_data_mask(result_data, mask_args[0])
+        return args[0]
+    else:
+        result_mask = _get_at_least_one_mask(*args[:2])
```
Comment (bhosmer) [inline torch/masked/maskedtensor/binary.py:160]: does this mean for non-inplace ops we're not sure if args[0] is a masked tensor?

File: torch/masked/maskedtensor/binary.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        # sparse tensors don't have strides
```
Comment (bhosmer) [inline torch/masked/maskedtensor/binary.py:161]: sorry to be dense... can you unpack how this relates to the code below?

**Review by bhosmer — COMMENTED** [review]
hey, first round of comments inline, mostly re the invariants and assumptions in `binary_helper`, putting it up to unblock and continuing 😁

File: torch/masked/maskedtensor/passthrough.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    torch.ops.aten.index,
+    torch.ops.aten.expand,
+    torch.ops.aten.view,
+    torch.ops.aten._unsafe_view,
+    torch.ops.aten._reshape_alias,
+    torch.ops.aten.cat,
+]
+
+
+# TODO: tree_map doesn't cut it due to kwargs support for torch_function
```
Comment (bhosmer) [inline torch/masked/maskedtensor/passthrough.py:31]: is this still true?

File: torch/masked/maskedtensor/passthrough.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        elif isinstance(a, list):
+            a_impl, _ = _map_mt_args_kwargs(a, {}, map_fn)
+            impl_args.append(a_impl)
+        elif isinstance(a, tuple):
+            a_impl, _ = _map_mt_args_kwargs(a, {}, map_fn)
+            impl_args.append(tuple(a_impl))
+        else:
+            impl_args.append(a)
+    impl_kwargs = {
+        k: map_fn(v) if is_masked_tensor(v) else v for (k, v) in kwargs.items()
```
Comment (bhosmer) [inline torch/masked/maskedtensor/passthrough.py:52]: above we're recursing into list and tuples for args, presumably we'd want to do the same for kwargs, no?

File: torch/masked/maskedtensor/passthrough.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    return impl_args, impl_kwargs
+
+
+def _wrap_result(result_data, result_mask):
+    if isinstance(result_data, list):
+        return list(_wrap_result(r, m) for (r, m) in zip(result_data, result_mask))
+    if isinstance(result_data, tuple):
+        return tuple(_wrap_result(r, m) for (r, m) in zip(result_data, result_mask))
+    if torch.is_tensor(result_data):
+        from .core import MaskedTensor
```
Comment (bhosmer) [inline torch/masked/maskedtensor/passthrough.py:63]: nit: hoist import

File: torch/masked/maskedtensor/passthrough.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+        return MaskedTensor(result_data, result_mask)
+    # Expect result_data and result_mask to be Tensors only
+    return NotImplemented
```
Comment (bhosmer) [inline torch/masked/maskedtensor/passthrough.py:67]: I'm actually not sure why `_wrap_result` and `_map_mt_args_kwargs` are in this file, it looks like they're used by binary.py and unary.py as well. maybe they belong in core.py, if that's where the common utils are?

File: torch/masked/maskedtensor/passthrough.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+def _is_pass_through_fn(fn):
+    return fn in PASSTHROUGH_FNS
+
+
+def _apply_pass_through_fn(fn, *args, **kwargs):
+    data_args, data_kwargs = _map_mt_args_kwargs(args, kwargs, lambda x: x.get_data())
+    result_data = fn(*data_args, **data_kwargs)
+    mask_args, mask_kwargs = _map_mt_args_kwargs(args, kwargs, lambda x: x.get_mask())
+    result_mask = fn(*mask_args, **mask_kwargs)
```
Comment (bhosmer) [inline torch/masked/maskedtensor/passthrough.py:78]: IIUC `_map_mt_args_kwargs()` will always return the original argument in the case of ordinary Tensors, disregarding the lambda. What does that mean is happening here when a mix of masked and ordinary tensors is passed?

File: torch/masked/maskedtensor/unary.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    "polygamma",
+    "real",
+    "remainder",
+    "true_divide",
+    "xlogy",
+]
+
+
+def _unary_helper(fn, args, kwargs, inplace):
+    from .passthrough import _map_mt_args_kwargs, _wrap_result
```
Comment (bhosmer) [inline torch/masked/maskedtensor/unary.py:108]: nit: hoist import

File: torch/masked/maskedtensor/unary.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+    if len(kwargs) != 0:
+        raise ValueError("MaskedTensor unary ops require that len(kwargs) != 0. "
```
Comment (bhosmer) [inline torch/masked/maskedtensor/unary.py:111]: s/!=/==/

File: torch/masked/maskedtensor/unary.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+                         "If you need support for this, please open an issue on Github.")
+    if len(args) > 1:
+        for a in args[1:]:
+            if torch.is_tensor(a):
+                raise TypeError("MaskedTensor unary ops do not support Tensor arguments")
```
Comment (bhosmer) [inline torch/masked/maskedtensor/unary.py:116]: what's an example callsite where we would perform this check?

File: torch/masked/maskedtensor/unary.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+    data_args, data_kwargs = _map_mt_args_kwargs(
+        args, kwargs, lambda x: x._masked_data
+    )
+
+    if args[0].layout() == torch.sparse_coo:
+        data_args[0] = data_args[0].coalesce()
+        s = data_args[0].size()
+        i = data_args[0].indices()
+        data_args[0] = data_args[0].coalesce().values()
+        v = fn(*data_args)
```
Comment (bhosmer) [inline torch/masked/maskedtensor/unary.py:130]: the prep work here seems to assume that len(args) == 1, but the invariant checking allows > 1 (and does some additional checks in that case).

File: torch/masked/maskedtensor/unary.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+        result_data = torch.sparse_coo_tensor(i, v, size=s)
+
+    elif args[0].layout() == torch.sparse_csr:
+        crow = data_args[0].crow_indices()
+        col = data_args[0].col_indices()
+        data_args[0] = data_args[0].values()
+        v = fn(*data_args)
```
Comment (bhosmer) [inline torch/masked/maskedtensor/unary.py:137]: ditto

**Review by bhosmer — COMMENTED** [review]
comments round 2/2 😁  definitely happy to chat with higher bandwidth offline if that's useful!

File: torch/masked/maskedtensor/binary.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+import torch
+
+from .core import _masks_match, _tensors_match, is_masked_tensor
+
+__all__ = []  # type: ignore[var-annotated]
+
+BINARY_NAMES = [
+    "add",
+    #    "addcdiv",
+    #    "addcmul",
```
[the remark this answers is not in the document]
Comment (george-qi) [reply torch/masked/maskedtensor/binary.py:12]: Probably easier to just remove for now!

File: torch/masked/maskedtensor/binary.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+        raise TypeError("At least one of `a` and `b` must be a MaskedTensor")
+    if not _masks_match(a, b):
+        raise ValueError("a and b must have matching masks")
+    if is_masked_tensor(a):
+        return a.get_mask()
+    return b.get_mask()
+
+
+def _binary_helper(fn, args, kwargs, inplace):
+    from .passthrough import _map_mt_args_kwargs, _wrap_result
```
[the remark this answers is not in the document]
Comment (george-qi) [reply torch/masked/maskedtensor/binary.py:85]: No real reason, will just move above

File: torch/masked/maskedtensor/binary.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+
+    if len(kwargs) != 0:
+        raise ValueError("len(kwargs) must equal 0")
+    if len(args) > 2:
+        for a in args[1:]:
+            if torch.is_tensor(a):
+                raise TypeError("MaskedTensor binary ops do not support Tensor arguments")
```
[the remark this answers is not in the document]
Comment (george-qi) [reply torch/masked/maskedtensor/binary.py:92]: Ahhh good call, I think you're right. I think `args[1:]` should just be `args[2:]` -- that is, we don't support Tensor arguments beyond the LHS and RHS for binary ops. Does that seem more reasonable?

File: torch/masked/maskedtensor/binary.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (bhosmer) [reply torch/masked/maskedtensor/binary.py:12]: well, it would add information to leave them commented and say why - like, are they just TODOs? or are they blocked on some upstream functionality that's not there yet? etc.

File: torch/masked/maskedtensor/binary.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (bhosmer) [reply torch/masked/maskedtensor/binary.py:92]: It seems more reasonable to me, but you're the authority 😛  if that's the correct invariant, i'd
- comment it
- adjust the error message
- remove the if, since it's redundant with `args[2:]`
(re this last, same could be done for the `if len(args) > 1' / `for args[1:]` in the unary_helper, IIRC

File: torch/masked/maskedtensor/binary.py [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+
+    if not _masks_match(*args[:2]):
+        raise ValueError(
+            "Input masks must match. If you need support for this, please open an issue on Github."
+        )
+
+    if (
+        is_masked_tensor(args[0])
+        and is_masked_tensor(args[1])
+        and args[0].layout() != args[1].layout()
```
[the remark this answers is not in the document]
Comment (bhosmer) [reply torch/masked/maskedtensor/binary.py:102]: yeah I think that's what the code below is relying on

[the remark this answers is not in the document]
Comment (george-qi) [reply torch/masked/maskedtensor/binary.py:102]: Sorry deleted my comment but you responded before I revised it! The code is meant to allow for lhs: MaskedTensor, rhs: dense tensor/float. The `is_masked_tensor` check at line 127 should read "if it has the same format"

File: torch/masked/maskedtensor/binary.py [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+                    "sparse_coo indices must match. If you need support for this, please open an issue on Github."
+                )
+            if data_args[0].size() != data_args[1].size():
+                raise ValueError("input1 and input2 must have the same size for binary functions.")
+
+        i = data_args[0].indices()
+        size = data_args[0].size()
+        data_args[0] = data_args[0].values()
+        if is_masked_tensor(args[1]):
+            data_args[1] = data_args[1].values()
```
[the remark this answers is not in the document]
Comment (george-qi) [reply torch/masked/maskedtensor/binary.py:128]: Changing this to only apply if `args[0]` and `args[1]` have the same layout

File: torch/masked/maskedtensor/binary.py (hunk 1 above)
Comment (george-qi) [reply torch/masked/maskedtensor/binary.py:160]: Yeah that's right

File: torch/masked/maskedtensor/binary.py (hunk 2 above)
Comment (george-qi) [reply torch/masked/maskedtensor/binary.py:161]: added some more detail in a comment

File: torch/masked/maskedtensor/passthrough.py (hunk 4 above)
Comment (george-qi) [reply torch/masked/maskedtensor/passthrough.py:52]: Yeah that makes sense, fixed

File: torch/masked/maskedtensor/passthrough.py (hunk 6 above)
Comment (george-qi) [reply torch/masked/maskedtensor/passthrough.py:67]: Ahhh that's true.. okay moving!

File: torch/masked/maskedtensor/unary.py (hunk 11 above)
Comment (george-qi) [reply torch/masked/maskedtensor/unary.py:130]: Sorry I don't quite follow - where does it assume that `len(args) == 1`?

File: torch/masked/maskedtensor/binary.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (george-qi) [reply torch/masked/maskedtensor/binary.py:92]: Yep that sounds great :)

File: torch/masked/maskedtensor/unary.py (hunk 10 above)
Comment (george-qi) [reply torch/masked/maskedtensor/unary.py:116]: We don't have much negative testing right now - I can add more of that in a follow up diff. Noting that as a to do!

File: torch/masked/maskedtensor/passthrough.py (hunk 3 above)
Comment (george-qi) [reply torch/masked/maskedtensor/passthrough.py]: This is a comment from Christian - I'll remove it for now and follow up at some point

File: torch/masked/maskedtensor/binary.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (george-qi) [reply torch/masked/maskedtensor/binary.py]: Yeah I believe we ran in to some edge cases with them and decided that they weren't high priority binary functions, so we tabled them for the time being. I'll add these as to do's as well

File: torch/masked/maskedtensor/binary.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (bhosmer) [reply torch/masked/maskedtensor/binary.py:92]: I think 127 moved since you commented :) also though, I'd suggest writing out the exact set of allowed layout combinations in a comment, and merging/deduplicating the logic here and in 105-109 into a single block. That'll make it much easier to understand/check the invariants that are being enforced here.

File: torch/masked/maskedtensor/binary.py [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+    )
+    mask_args, mask_kwargs = _map_mt_args_kwargs(
+        args, kwargs, lambda x: x.get_mask()
+    )
+
+    args0_layout = args[0].layout() if is_masked_tensor(args[0]) else args[0].layout
+    same_layout = (
+        (is_masked_tensor(args[1]) and args[1].layout() != args0_layout)
+        or (torch.is_tensor(args[1]) and args[1].layout != args0_layout)
+    )
```
Comment (bhosmer) [inline torch/masked/maskedtensor/binary.py:109]: not sure why this is called `same_layout`? the condition it seems to be testing is basically the opposite, no?

Comment (bhosmer) [reply torch/masked/maskedtensor/binary.py:109]: also you should be able to dispense with all the `is_masked_tensor()` branching here and just use `data_args` instead of `args`

File: torch/masked/maskedtensor/unary.py (hunk 11 above)
Comment (bhosmer) [reply torch/masked/maskedtensor/unary.py:129]: eh nvm, I must've jumped to thinking a masked tensor wouldn't trigger the `is_tensor` check on 114, my mistake

**Review by bhosmer — COMMENTED** [review]
looking good, thanks for the quick turnaround! I'd say the invariant checks at the top of binary_fn_helper will definitely benefit from some further rationalization, but other than that LGTM 😁

File: torch/masked/maskedtensor/passthrough.py (hunk 7 above)
Comment (bhosmer) [reply torch/masked/maskedtensor/passthrough.py:41]: right, but doesn't that mean we're sometimes doing redundant computation when one or the other argument is an ordinary tensor? (given that at least one is always a MT)

File: torch/masked/maskedtensor/binary.py (hunk 18 above)
Comment (george-qi) [reply torch/masked/maskedtensor/binary.py:109]: Sorry yeah it should be `==` instead of `!=`. And yeah that's fair, I suppose at the end of the day I only care about the `data_args`

Comment (george-qi) [reply torch/masked/maskedtensor/binary.py]: Regarding the 105-109 logic from above -- I think it'd be best to just remove that for now and allow for different layouts since that's the end goal anyway

Comment (bhosmer) [reply torch/masked/maskedtensor/binary.py]: just curious, I'd have expected tests to catch the !=/== issue above, are we testing all combinations of layouts? I thought we were via parameterize but maybe I misread - we should be, though

**Review by bhosmer — APPROVED** [review]
LG, thanks for the changes!
